jacobian_follower/src/jacobian_follower/jacobian_follower.cpp

```diff
@@ -361,50 +408,89 @@ PlanResult JacobianFollower::moveInWorldFrame(planning_scene_monitor::LockedPlan
     for (auto ee_idx = 0ul; ee_idx < num_ees; ++ee_idx) {
-      robotTtargets[ee_idx].translation() = robot_to_world * tool_paths[ee_idx][step_idx];
+      robotTtargets[ee_idx].translation() = robot_to_world * tools_waypoint_interpolated[ee_idx][step_idx];
       robotTtargets[ee_idx].linear() = preferred_tool_orientations[ee_idx].toRotationMatrix();
     }
 
     // Note that jacobianIK is assumed to have modified the state in the planning scene
-    const auto iksoln = jacobianIK(planning_scene, world_to_robot, jmg, tool_names, robotTtargets);
+    const auto iksoln = jacobianIK(planning_scene, world_to_robot, jmg, tool_names, robotTtargets, constraint_fun_);
     if (!iksoln) {
       ROS_DEBUG_STREAM_NAMED(LOGGER_NAME, "IK Stalled at idx " << step_idx << ", returning early");
       break;
     }
     auto const state = planning_scene->getCurrentState();
     ROS_DEBUG_STREAM_NAMED(LOGGER_NAME, "step_idx " << step_idx);
+    std::stringstream tool_pos_ss;
+    for (auto ee_idx = 0ul; ee_idx < num_ees; ++ee_idx) {
+      auto ee_pos = robotTtargets[ee_idx].translation();
+      tool_pos_ss << ee_pos(0) << "," << ee_pos(1) << "," << ee_pos(2) << " ";
+    }
+    ROS_DEBUG_STREAM_NAMED(LOGGER_NAME + ".tool_positions", "tool positions " << tool_pos_ss.str());
     debugLogState("jacobianIK: ", state);
-    cmd.addSuffixWayPoint(state, 0);
+    trajectory.addSuffixWayPoint(state, 0);
   }
 
-  ROS_DEBUG_NAMED(LOGGER_NAME, "Jacobian IK path has %lu points out of a requested %lu", cmd.getWayPointCount(),
+  ROS_DEBUG_NAMED(LOGGER_NAME, "Jacobian IK path has %lu points out of a requested %lu", trajectory.getWayPointCount(),
                   steps - 1);
-  return cmd;
+  return trajectory;
+}
+
+bool JacobianFollower::checkCollision(planning_scene::PlanningScenePtr planning_scene,
+                                      robot_state::RobotState const &state) {
+  collision_detection::CollisionRequest collisionRequest;
+  collisionRequest.contacts = true;
+  collisionRequest.max_contacts = 1;
+  collisionRequest.max_contacts_per_pair = 1;
+  collision_detection::CollisionResult collisionResult;
+  planning_scene->checkCollision(collisionRequest, collisionResult, state);
+  if (collisionResult.collision) {
+    ROS_DEBUG_STREAM_NAMED(LOGGER_NAME, "Collision Result: " << collisionResult);
+    //    std::stringstream ss;
+    //    planning_scene->getAllowedCollisionMatrix().print(ss);
+    //    ROS_DEBUG_STREAM_NAMED(LOGGER_NAME, "\n" << ss.str());
+
+    //    std::vector<moveit_msgs::AttachedCollisionObject> msgs;
+    //    planning_scene->getAttachedCollisionObjectMsgs(msgs);
+    //    for (auto const &m : msgs)
+    //    {
+    //      ROS_DEBUG_STREAM_NAMED(LOGGER_NAME, "Touch Links: " << m.link_name << " " << m.touch_links);
+    //    }
+
+    if (visualize_) {
+      //      visual_tools_.publishContactPoints(collisionResult.contacts, planning_scene.get());
+      visual_tools_.publishRobotState(planning_scene->getCurrentState());
```

Comment:
you can visualize the non-collision case after you get the returned trajectory. The reason you need viz here when there is collision is because the in-collision state doesn't get returned

Reply:
I see. The problem is you end up visualizing one "time-step" in very different places of code, depending on if there is a collision or not. You might change your visualization for your collision free, then only later have to track down the collision visualization deep in jacobian_follower. I have a ton of this in my code and I don't like it

Anyway, fine for now but it is annoying in the long run